cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,172 @@
+func RunApply(flags *applyFlags) error {
+
+	// Start with the basics, verify that the cluster is healthy and get config
+	client, cfg, err := EnforceRequirements(flags.parent.kubeConfigPath, flags.parent.cfgPath, flags.parent.printConfig)
+	if err != nil {
+		return err
+	}
+
+	// Set the version on the config object now
+	cfg.KubernetesVersion = flags.newK8sVersionStr
+
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	api.Scheme.Convert(cfg, internalcfg, nil)
+
+	fmt.Printf("[upgrade/version] You have chosen to upgrade to version %q\n", flags.newK8sVersionStr)
+	versionGetterImpl := upgrade.NewKubeVersionGetter(client, os.Stdout)
+	versionSkewErrs := upgrade.EnforceVersionPolicies(versionGetterImpl, flags.newK8sVersionStr, flags.newK8sVersion, flags.parent.allowExperimentalUpgrades, flags.parent.allowRCUpgrades)
+	if versionSkewErrs != nil {
+
+		if len(versionSkewErrs.Mandatory) > 0 {
+			fmt.Printf("[upgrade/version] FATAL: The --version argument is invalid: %v.\n[upgrade/version] Can be bypassed if you pass the --force flag\n", versionSkewErrs.Mandatory)
+			return fmt.Errorf("The --version argument is invalid")
+		}
+
+		if len(versionSkewErrs.Skippable) > 0 {
+			fmt.Printf("[upgrade/version] ERROR: The --version argument is invalid: %v.\n[upgrade/version] Can be bypassed if you pass the --force flag\n", versionSkewErrs.Skippable)
+
+			if !flags.force {
+				return fmt.Errorf("The --version argument is invalid")
+			}
+			fmt.Println("[upgrade/version] Continuing since the --force flag is set")
+		}
+	}
+
+	if !flags.flagYes && !flags.dryRun {
+
+		fmt.Print("[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: ")
+
+		scanner := bufio.NewScanner(os.Stdin)
+		scanner.Scan()
+		if err = scanner.Err(); err != nil {
+			return fmt.Errorf("couldn't read from standard input: %v", err)
+		}
+		answer := scanner.Text()
+		if strings.ToLower(answer) != "y" {
+			return fmt.Errorf("won't proceed; the user didn't answer (Y|y) in order to continue")
+		}
+	}
+
+	// Use a prepuller implementation based on creating DaemonSets
+	prepuller := upgrade.NewDaemonSetPrepuller(client, internalcfg)
+	// Block until all DaemonSets are ready
+	// TODO: Make this timeout configurable?
+	upgrade.PrepullImagesInParallell(prepuller, 15*time.Minute)
+
+	if upgrade.IsControlPlaneSelfHosted(client) {
+		fmt.Printf("[upgrade/apply] Upgrading your Self-Hosted control plane to version %q...\n", flags.newK8sVersionStr)
+
+		// Upgrade a self-hosted cluster
+		return fmt.Errorf("not implemented")
+	} else {
+		fmt.Printf("[upgrade/apply] Upgrading your Static Pod-hosted control plane to version %q...\n", flags.newK8sVersionStr)
+
+		// Upgrade a static-pod hosted cluster
+		if err := upgrade.UpgradeStaticPodControlPlane(client, internalcfg, flags.newK8sVersion); err != nil {
+			return err
+		}
+
+		// Upgrade the static pod hosted control plane to a self-hosted one
+		// TODO: Do this only in case the self-hosting feature gate is enabled
```

Comment:
now fixed; we won't upgrade a static pod cluster to a self-hosted one in v1.8, so I could just remove this